_posts/2015-03-29-organizing-data-self-encapsulation.markdown

```diff
@@ -0,0 +1,222 @@
+
+Self Encapsulation offers a surprising number of advantages, though. What if we introduce a sub-class of Person to represent medical doctors. With any of the examples above we could implement a Doctor's first name as:
+
+{% highlight ruby %}
+class Doctor < Person
+  def full_name
+    "Dr. #{super}"
+  end
+end
+
+Doctor.new("Marcus", "Welby").full_name
+#=> "Dr. Marcus Welby"
+{% endhighlight %}
+
+So far, so good.
+
+Then, later, you add a getter for `first_name` in Person. Maybe you have to work around some capitalization problems.
+
+{% highlight ruby %}
+class Person
+  def first_name
+    @first_name.capitalize
+  end
+end
+{% endhighlight %}
+
+Now your direct access code looks like this:
+
+{% highlight ruby %}
+class Person
+  def initialize(first_name, last_name)
+    @first_name = first_name
+    @last_name = last_name
+  end
+
+  def full_name
+    "#{@first_name} #{@last_name}"
+  end
+
+  def first_name
+    @first_name.capitalize
+  end
+end
+
+class Doctor < Person
+  def full_name
+    "Dr. #{super}"
+  end
+end
+
+Doctor.new("marcus", "Welby").full_name
+#=> "Dr. marcus Welby"
+{% endhighlight %}
+
+Because you are human you have created a `first_name` getter and yet you are still using the instance variable `@first_name` in `full_name`. Their implementations differ and you find yourself in Bug Town. Note that I'm making a big assumption here: that you are human. I know! This might be crazy. But play along with me. You didn't create this bug this because you are dumb, or because you are a bad programmer, this is just a simple mistake that humans make.
+
+The Mixed Variable Access method protects you from this bit of humanity. Since you're using the getter method from the beginning, you don't have to worry about introducing a getter later. After implementing Doctor, our Mixed Access code looks like this:
+
+{% highlight ruby %}
+class Person
+  attr_reader :last_name
+
+  def initialize(first_name, last_name)
+    @first_name = first_name
+    @last_name = last_name
+  end
+
+  def full_name
+    "#{first_name} #{last_name}"
+  end
+
+  def first_name
+    @first_name.capitalize
+  end
+end
+
+class Doctor < Person
+  def full_name
+    "Dr. #{super}"
+  end
+end
+
+Doctor.new("marcus", "Welby").full_name
+#=> "Dr. Marcus Welby"
+{% endhighlight %}
+
+As your code deals with more and more people you find more problems. Doctors have taken to entering their last name in the database with the suffix, ", MD" so that it gets displayed an "Dr. Marcus Welby, MD". Your bosses decide all Doctors should have their names displayed in this format.
+
+So, you fix the problem by changing how the full name is displayed and forcibly removing the now-unnecessary input.
+
+{% highlight ruby %}
+class Doctor < Person
+  def last_name=(x)
+    @last_name = x.delete(", MD")
+  end
```

Comment:
You're adding a setter here where there weren't any before. SInce the mixed access scenario doesn't use any setters, it makes more sense I think to solve this in the initializer, e.g.:

```ruby
class Doctor < Person
  def initialize(first_name, last_name)
    super first_name, last_name.delete(", MD")
  end
end
```